src/app/stream-deck/actions.ts

```diff
@@ -0,0 +1,379 @@
+import { showNotification } from 'app/notifications/notifications';
+import { setSearchQuery } from 'app/shell/actions';
+import { refresh } from 'app/shell/refresh-events';
+import { RootState, ThunkResult } from 'app/store/types';
+import {
+  streamDeckMaxPowerUpdate,
+  streamDeckMetricsUpdate,
+  streamDeckPostMasterUpdate,
+  streamDeckVaultUpdate,
+} from 'app/stream-deck/stream-deck-update';
+import { observeStore } from 'app/utils/redux-utils';
+import _ from 'lodash';
+import { createAction } from 'typesafe-actions';
+
+let streamDeckWebSocket: WebSocket;
+
+let refreshInterval: ReturnType<typeof setInterval>;
+
+// Cache shares to loadouts weakly, to cut down on creating shares
+const loadoutShares = new WeakMap<Loadout, string>();
+
+export const streamDeckConnected = createAction('stream-deck/CONNECTED')();
+
+export const streamDeckDisconnected = createAction('stream-deck/DISCONNECTED')();
+
+export const streamDeckChangeStatus = createAction('stream-deck/CHANGE-STATUS')<boolean>();
+
+export const streamDeckWaitSelection = createAction('stream-deck/WAIT-SELECTION')<
+  'loadout' | 'item'
+>();
+
+export const streamDeckClearSelection = createAction('stream-deck/CLEAR-SELECTION')();
+
+interface StreamDeckMessage {
+  action:
+    | 'search'
+    | 'randomize'
+    | 'collectPostmaster'
+    | 'refresh'
+    | 'farmingMode'
+    | 'maxPower'
+    | 'freeSlot'
+    | 'pullItem'
+    | 'selection'
+    | 'loadout'
+    | 'shareLoadout';
+  args: {
+    search: string;
+    weaponsOnly: boolean;
+    loadout: string;
+    character: string;
+    slot: InventoryBucket['type'];
+    item: string;
+    page: string;
+    selection: 'loadout' | 'item';
+  };
+}
+
+// serialize the data and send it if connected
+export function sendToStreamDeck(args: Record<string, any>): ThunkResult {
+  return async () => {
+    if (streamDeckWebSocket?.readyState === WebSocket.OPEN) {
+      streamDeckWebSocket.send(JSON.stringify(args));
+    }
+  };
+}
+
+// on click on InventoryItem send the selected item to the Stream Deck
+export function streamDeckSelectItem(item: DimItem): ThunkResult {
+  return async (dispatch, getState) => {
+    const { streamDeck } = getState();
+    if (streamDeck.enabled && streamDeck.selection === 'item') {
+      hideItemPopup();
+      streamDeck.selectionPromise.resolve();
+      dispatch(streamDeckClearSelection());
+      return dispatch(
+        sendToStreamDeck({
+          selectionType: 'item',
+          selection: {
+            label: item.name,
+            subtitle: item.typeName,
+            item: item.id,
+            icon: item.icon,
+          },
+        })
+      );
+    }
+  };
+}
+
+function findSubClass(items: LoadoutItem[], state: RootState) {
+  const defs = d2ManifestSelector(state);
+  for (const item of items) {
+    const def = defs?.InventoryItem.get(item.hash);
+    // find subclass item
+    if (def?.inventory?.bucketTypeHash === 3284755031) {
```

Comment:
`import { BucketHashes } from 'data/d2/generated-enums';`

```suggestion
    if (def?.inventory?.bucketTypeHash === BucketHashes.Subclass) {
```